File: app/frontend/src/lib/api/audit.ts
import { apiFetch } from "../api-client";

export interface AuditLog {
  id: string;
  actor_user_id: string | null;
  action: string;
  object_type: string;
  object_id: string | null;
  patient_id?: string | null;
  outcome: string;
  trace_id?: string | null;
  ip_address?: string | null;
  metadata?: Record<string, unknown> | null;
  created_at: string;
}

export interface AuditLogList {
  items: AuditLog[];
}

export interface AuditLogFilters {
  patient_id?: string;
  action?: string;
  outcome?: string;
  limit?: number;
}

export async function getAuditLogs(filters?: AuditLogFilters): Promise<AuditLogList> {
  const params = new URLSearchParams();
  if (filters?.patient_id) params.append("patient_id", filters.patient_id);
  if (filters?.action) params.append("action", filters.action);
  if (filters?.outcome) params.append("outcome", filters.outcome);
  if (filters?.limit) params.append("limit", filters.limit.toString());

  return apiFetch<AuditLogList>(`/audit/logs?${params.toString()}`);
}

File: app/backend/src/hospital_ai/services/audit.py
import uuid
from typing import Optional, Any

from sqlalchemy.ext.asyncio import AsyncSession

from hospital_ai.db.models import AuditLog


class AuditService:
    def __init__(self, session: AsyncSession) -> None:
        self.session = session

    async def record(
        self,
        *,
        actor_user_id: uuid.Optional[UUID],
        action: str,
        object_type: str,
        outcome: str,
        trace_id: str,
        object_id: uuid.Optional[UUID] = None,
        patient_id: uuid.Optional[UUID] = None,
        ip_address: Optional[str] = None,
        metadata: dict[str, Optional[Any]] = None,
    ) -> AuditLog:
        log = AuditLog(
            actor_user_id=actor_user_id,
            action=action,
            object_type=object_type,
            object_id=object_id,
            patient_id=patient_id,
            outcome=outcome,
            trace_id=trace_id,
            ip_address=ip_address,
            meta=metadata or {},
        )
        self.session.add(log)
        await self.session.flush()
        return log
